File: src/utils/environment.ts
import { cleanEnv, num, str } from "envalid";

interface Environment {
  port: number;
  l1Rpc: string;
  l2Rpc: string;
  l1ChainId: string;
  l2ChainId: string;
  l1Cdm: string;
  l2Cdm: string;
  l1CrossDomainMessengerAddress: string;
  l2OutputOracle: string;
  l1Mnt: string;
  l2Mnt: string;
  l2Eth: string;
  ethGasLimit: number;
  mntGasLimit: number;
}

export const environment: Environment = {
  port: 0,
  l1Rpc: "",
  l2Rpc: "",
  l1ChainId: "",
  l2ChainId: "",
  l1Cdm: "",
  l2Cdm: "",
  l1CrossDomainMessengerAddress: "",
  l2OutputOracle: "",
  l1Mnt: "",
  l2Mnt: "",
  l2Eth: "",
  ethGasLimit: 0,
  mntGasLimit: 0,
};

export const setupEnvironment = async () => {
  const devEnvs = cleanEnv(process.env, {
    PORT: num(),
    L1_RPC: str(),
    L2_RPC: str(),
    L1_CHAINID: str(),
    L2_CHAINID: str(),
    L1_CDM: str(),
    L2_CDM: str(),
    L1_CROSS_DOMAIN_MESSENGER_ADDRESS: str(),
    L2_OUTPUT_ORACLE: str(),
    L1_MNT: str(),
    L2_MNT: str(),
    L2_ETH: str(),
    ETH_GAS_LIMIT: num(),
    MNT_GAS_LIMIT: num(),
  });
  environment.port = devEnvs.PORT;
  environment.l1Rpc = devEnvs.L1_RPC;
  environment.l2Rpc = devEnvs.L2_RPC;
  environment.l1ChainId = devEnvs.L1_CHAINID;
  environment.l2ChainId = devEnvs.L2_CHAINID;
  environment.l1Cdm = devEnvs.L1_CDM;
  environment.l2Cdm = devEnvs.L2_CDM;
  environment.l1CrossDomainMessengerAddress =
    devEnvs.L1_CROSS_DOMAIN_MESSENGER_ADDRESS;
  environment.l2OutputOracle = devEnvs.L2_OUTPUT_ORACLE;
  environment.l1Mnt = devEnvs.L1_MNT;
  environment.l2Mnt = devEnvs.L2_MNT;
  environment.l2Eth = devEnvs.L2_ETH;
  environment.ethGasLimit = devEnvs.ETH_GAS_LIMIT;
  environment.mntGasLimit = devEnvs.MNT_GAS_LIMIT;
};

File: src/pkg/utils/environment.go
package utils

import (
	"fmt"
	"os"
)

// Environment represents the application's environment configuration
type Environment struct {
	Port        string
	L1RPC       string
	L2RPC       string
	L1ChainID   int64
	L2ChainID   int64
	L1MNT       string
	L2MNT       string
	L2Bridge    string
	L2Eth       string
	EthGasLimit int64
	MNTGasLimit int64
}

var env Environment

// SetupEnvironment initializes the environment variables
func SetupEnvironment() error {
	env = Environment{
		Port:        getEnv("PORT", "8080"),
		L1RPC:       getEnv("L1_RPC", ""),
		L2RPC:       getEnv("L2_RPC", ""),
		L1ChainID:   getEnvInt64("L1_CHAINID", 1),
		L2ChainID:   getEnvInt64("L2_CHAINID", 5000),
		L1MNT:       getEnv("L1_MNT", ""),
		L2MNT:       getEnv("L2_MNT", ""),
		L2Bridge:    getEnv("L2_BRIDGE", ""),
		L2Eth:       getEnv("L2_ETH", ""),
		EthGasLimit: getEnvInt64("ETH_GAS_LIMIT", 1000000),
		MNTGasLimit: getEnvInt64("MNT_GAS_LIMIT", 1000000),
	}
	return nil
}

func getEnv(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}

func getEnvInt64(key string, defaultValue int64) int64 {
	if value := os.Getenv(key); value != "" {
		var result int64
		_, err := fmt.Sscanf(value, "%d", &result)
		if err == nil {
			return result
		}
	}
	return defaultValue
}

// GetEnvironment returns the current environment configuration
func GetEnvironment() Environment {
	return env
}

// GetTestERC20ABI returns the TestERC20 ABI
func GetTestERC20ABI() string {
	return `[{"constant":true,"inputs":[{"name":"_owner","type":"address"}],"name":"balanceOf","outputs":[{"name":"balance","type":"uint256"}],"type":"function"}]`
}
